### PR TITLE
Cleanup merge. Removed changeset related to Device.cs.

### PR DESCRIPTION
**Pull request by BionicCode**
- Merge changes from #41 to branch iasyncenumerable
  - Target #39: Fixed LINQ expression to use the conditional `Enumerable.FirstOrDefault` instead of `Enumerable.First` in `DeviceLocator.GetMulticastIPAddressesForDiscovery`. `DeviceLocator.GetMulticastIPAddressesForDiscovery` now returns an empty `IEnumerable` in case the condition is never satisfied.

  - Improve API: Changed `DiscoverAsync` to return a `IEnumerable<Device>` instead of a `List<T`. Method signature remains unchanged. Basically removed call to `Enumerable.ToList` to defer the execution of the query in order to improve performance (especially in situations where the caller wants to iterate over/filter the result collection).

  -  Add cancellation support to asynchronous API by providing corresponding overloads (non-breaking change)

  - Update old XML documantation

  - Add/improve XML documantation

  - Run Resharper code cleanup (fix indention and formatting).

- Fix/Implement `IAsyncEnumerable EnumerateDevicesAsync(void)` and `IAsyncEnumerable EnumerateDevicesAsync(CancellationToken)`
- Implement `EnumerateDevicesAsync` overload which accepts `NetworkInterface`. 
- Fix test case after renaming method to `EnumerateDevicesAsync`. 
- Add `EnumeratorCancellation` attribute to `IAsyncEnumerable` methods. 
- Refactored `DeviceLocator` device lookup (`GetDevicesFromSocket` in particular) to support enumeration (may improve performance).
- Add throw `DeviceDeiscoveryException` when a `SocketException` was thrown and no device was found (optimistic exception handling - if exception was thrown but at least one device was discovered no exception will be thrown). 
- Add `DeviceDiscoveryException` as major excpetion of `DeviceLocator` and as a wrapper of any `SocketException` that was thrown during the discovery."

### CODE REVIEW

**Comment by roddone** [comment]
@BionicCode i've just ran the unit tests and there is a regression introduced by the new `DeviceDiscoveryException` when using the parameter "UseAllAvailableMulticastAddresses" with true value.
Enabling this parameter makes the API throw SocketExceptions when iterating over multicast addresses that are not reachable. This is normal and is not a bug because the parameter is designed to help people who use the API on Xamarin projects.

The consequence is the discovery that fails because when the exception is thrown for any of the unicast addresses which does not find devices, it breaks all the other Tasks in `SearchNetworkForDevicesAsync`.

**Comment by BionicCode** [comment]
@roddone  I see. The idea was to provide users that run into an issue like #39 some feedback to help to debug their environment. In this case the user was receiving zero devices although he expected to receive two. The reason might be pointed out by the exception message that was probably thrown. Every `SocketException` is currently silenced. Do you think there is a more graceful solution to test if a multicast address is valid? What do you think about this solution:  

When no device was found AND a `SocketException` was thrown we throw a `DeviceDiscoveryException` that accumulates all thrown exceptions (of each failed socket). The user can handle this exception or use the exception messages to debug his environment. (This is how it is right now). This is not the optimal solution as exceptions still get silenced in case at least one device was successfully discovered.

Currently `DiscoverAsync` iterates over all network interfaces returned from `NetworkInterface.GetAllNetworkInterfaces`. The first failed network interface will result in a `DeviceDiscoveryException` if a socket error occurred. Currently a network interface is failed when all its multicast addresses didn't connect to a device and a `SocketException was thrown. I collect all `DeviceDiscoveryException` on the most outer loop. If the above described condition is satisfied, I merge all `DeviceDiscoveryException` and re-throw it. If the user wants to ignore the exception, he must explicitly catch and swallow it.

Please let me know if you agree or if you want me to remove the exception based on a failed discovery due to failed sockets.

**Comment by BionicCode** [comment]
@roddone Is there a reason not to use the async API of `Socket`? I would like to replace e.g. `Socket.SendTo` with `Socket.SendToAsync`. Just use the async API where possible. Are you fine with it?
.

**Comment by BionicCode** [comment]
@roddone  I have added a new commit that improves the exception handling. Can you please verify this version?

**Comment by roddone** [comment]
@BionicCode The error was still happening, so I have done some changes to allow discovery to get all the SocketExceptions from all multicast adresses and to throw DeviceDiscoveryException in the end, in the case where no devices are found and socketExceptions has ocurred.
Can you have a look and tell me what you think about this ? (been in a rush, I probably haven't done the cleanest code I should have do)

File: YeelightAPI/DeviceLocator.cs [hunk 1]
```diff
@@ -20,877 +20,881 @@
...
+        /// <returns>A <see cref="Task" /> that contains a collection of all discovered <see cref="Device" /> instances.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when operation was cancelled by the caller.</exception>
+        public static async Task<IEnumerable<Device>> DiscoverAsync(
+          IProgress<Device> deviceFoundReporter,
+          CancellationToken cancellationToken)
+        {
+            IEnumerable<Task<DiscoveryResult>> tasks = NetworkInterface.GetAllNetworkInterfaces()
+              .Where(networkInterface => networkInterface.OperationalStatus == OperationalStatus.Up)
+              .Select(
+                networkInterface => DeviceLocator.SearchNetworkForDevicesAsync(networkInterface, deviceFoundReporter, cancellationToken));
```
Comment (BionicCode) [inline YeelightAPI/DeviceLocator.cs]: I think the call to `DeviceLocator.SearchNetworkForDevicesAsync` should be awaited for proper exception behavior. I missed this before. I have just realized this.

```
// Correct: async lambda expression
.Select(async networkInterface => await DeviceLocator.SearchNetworkForDevicesAsync(networkInterface, deviceFoundReporter, cancellationToken));
```

**Review by BionicCode — COMMENTED** [review]
Looks good. Found one issue regarding a lambda expression that should've bee async. I was missing this before.

**Comment by BionicCode** [comment]
@roddone Nice fix. I was not able to test it. Generally the exception handling should have worked before. But your solution looks cleaner as you are returning a simple result object. Have you already tested the `EnnumerateDevicesAsync` members? I have just implemented the exception handling following your pattern using the `DiscoveryResult`. I also have fixed the asnyc lamda and fixed the condition when to throw the exception. Please take a look at this (Line 307) when reviewing the changes.

Must be: _"throw DeviceDiscoveryException in the end, in the case where no devices are found and socketExceptions has ocurred."_

Found condition (Line 307):

```
// Will throw if any result has no devices AND if any (other) result has an error.
// Will discard all valid results e.g. in case when a result has an error but also a result (valid) AND some other result has no devices but also no error (valid).
if (!result.Any(r => r.Devices.Any()) && result.Any(r => r.Exceptions.Any())
```

Fixed condition:

```
// Will throw only when all results are failed => exception has occurred and no device was found
if (results.All(r => !result.HasResult && result.HasError))
```

File: YeelightAPI/DeviceLocator.cs [hunk 2]
```diff
@@ -301,15 +301,15 @@ private static List<Task<List<Device>>> CreateDiscoverTasks(NetworkInterface net
               .Select(
                 async networkInterface => await DeviceLocator.SearchNetworkForDevicesAsync(networkInterface, deviceFoundReporter, cancellationToken));
 
-            DiscoveryResult[] result = await Task.WhenAll(tasks);
+            DiscoveryResult[] results = await Task.WhenAll(tasks);
 
             //no results and exceptions has occured
-            if (result.All(r => !r.HasResult && r.HasError))
+            if (results.All(result => !result.HasResult && result.HasError))
```
Comment (BionicCode) [inline YeelightAPI/DeviceLocator.cs:120]: @roddone Please approve if this new condition is correct (or if the original condition is _really_ wrong as I remarked). The original condition was discarding valid results.

Expected condition: throw if no devices were found and an exception was thrown.

**Comment by roddone** [comment]
Looks good, As I said I was in a rush and had no time to check every cases.

I just released 1.9 which includes the music mode, I will release this PR as 1.10 with the followings : 
* remove deprecated methods
* loooots of unit tests around discovery, to ensure all optional parameters and all error cases are tested

There is no rush releasing this so I will try to do that this week or when I will be back from holidays in a few weeks ;)

**Comment by BionicCode** [comment]
@roddone  Sounds good 👍 Then have some nice holidays and stay healthy 😃

**Comment by nathanpovo** [comment]
Will these changes eventually be applied on top of the `master` branch?

This pull request is currently deviated from the `master` resulting in pull requests #42 and #45 being ignored.

image

**Comment by roddone** [comment]
Of course they will be merged into master as soon as I have time and solved my networks issues at home (just moved in another place, I haven't plugged back all my smart home stuff).

In the meantime, you can create a PR from your own fork to this branch `iasyncenumerable` so I can include it in my tests later ;)

**Pull request merged**